sdk/core/core/src/az_json_read.c

```diff
@@ -0,0 +1,379 @@
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_is_esc(char const c) {
+  switch (c) {
+    case '\\':
+    case '"':
+    case '/':
+    case 'b':
+    case 'f':
+    case 'n':
+    case 'r':
+    case 't':
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_stack_is_empty(az_json_state const *const p) {
+  return p->stack == 1;
+}
+
+inline az_json_stack_item az_json_stack_last(az_json_state const *const p) {
+  return p->stack & 1;
+}
+
+az_json_state az_json_state_create(az_const_str const buffer) {
+  return (az_json_state){
+    .buffer = buffer,
+    .i = 0,
+    .stack = 1,
+  };
+}
+
+az_result az_json_get_char(az_const_str const buffer, size_t const i, char *const out_char) {
+  if (i == buffer.size) {
+    return AZ_JSON_ERROR_UNEXPECTED_END;
+  }
+  *out_char = az_const_str_item(buffer, i);
+  return AZ_OK;
+}
+
+az_result az_json_expect_char(az_const_str const buffer, size_t *const p, char const expected) {
+  char c;
+  AZ_RETURN_ON_ERROR(az_json_get_char(buffer, *p, &c));
+  if (c != expected) {
+    return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+  }
+  *p += 1;
+  return AZ_OK;
+}
+
+az_result az_json_read_white_space(az_const_str const buffer, size_t *const p) {
+  for (; *p != buffer.size; ++*p) {
+    if (!az_json_is_white_space(az_const_str_item(buffer, *p))) {
+      break;
+    }
+  }
+  return AZ_OK;
+}
+
+az_result az_json_stack_push(az_json_state *const p_state, az_json_stack const stack) {
+  if (p_state->stack >> AZ_JSON_STACK_SIZE != 0) {
+    return AZ_JSON_ERROR_STACK_OVERFLOW;
+  }
+  p_state->stack = (p_state->stack << 1) | stack;
+  return AZ_OK;
+}
+
+az_result az_json_read_keyword_rest(
+  az_const_str const buffer,
+  size_t *const p,
+  az_const_str const keyword
+) {
+  *p += 1;
+  for (size_t k = 0; k != keyword.size; ++*p, ++k) {
+    if (*p == buffer.size) {
+      return AZ_JSON_ERROR_UNEXPECTED_END;
+    }
+    if (az_const_str_item(buffer, *p) != az_const_str_item(keyword, k)) {
+      return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+    }
+  }
+  return AZ_OK;
+}
+
+az_result az_json_read_number_digit_rest(
+  az_const_str const buffer,
+  size_t *const p,
+  double *const out_value,
+  int const sign
+) {
+  double value = 0;
+
+  // integer part
+  {
+    char c = az_const_str_item(buffer, *p);
+    if (c != '0') {
+      // read an integer part of the number
+      do {
+        value = value * 10 + (c - '0');
+        *p += 1;
+        if (*p == buffer.size) {
+          break;
+        }
+        c = az_const_str_item(buffer, *p);
+      } while (az_is_digit(c));
+    } else {
+      *p += 1;
+    }
+  }
+
+  // fraction
+  int16_t exp = 0;
+  if (*p != buffer.size && az_const_str_item(buffer, *p) == '.') {
+    *p += 1;
+    if (*p == buffer.size) {
+      return AZ_JSON_ERROR_UNEXPECTED_END;
+    }
+    char c = az_const_str_item(buffer, *p);
+    if (!az_is_digit(c)) {
+      return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+    }
+    do {
+      value = value * 10 + (c - '0');
```

Comment:
This is a bug
(shoutout to STL for coming up with this example)
consider:
```
#include <stdio.h>
using namespace std;
void print(const double d) {
    printf("%a %.0f\n", d, d);
}
int main() {
    print(9223372030000000000.0);
    puts("--");
    print(9223372030000000511.0);
    print(9223372030000000512.0);
    print(9223372030000000513.0);
    puts("--");
    print(9223372030000001024.0);
    puts("==========");
    print(922337203000000051.0 * 10.0 + 1.0);
    print(922337203000000051.0 * 10.0 + 2.0);
    print(922337203000000051.0 * 10.0 + 3.0);
}
C:\Temp>cl /EHsc /nologo /W4 charlie.cpp && charlie
charlie.cpp
0x1.fffffff99db1bp+62 9223372030000000000
--
0x1.fffffff99db1bp+62 9223372030000000000
0x1.fffffff99db1cp+62 9223372030000001024
0x1.fffffff99db1cp+62 9223372030000001024
--
0x1.fffffff99db1cp+62 9223372030000001024
==========
0x1.fffffff99db1bp+62 9223372030000000000
0x1.fffffff99db1bp+62 9223372030000000000
0x1.fffffff99db1bp+62 9223372030000000000
```

Reply:
there have been *many* bugs in the CRT because of implementing strtod like this.